src/server/scripts/EasternKingdoms/ZulGurub/boss_jeklik.cpp

```diff
@@ -265,45 +387,155 @@ struct boss_jeklik : public BossAI
+        CreatureAI::Reset();
+
+        switch (_mode)
+        {
+            case BAT_RIDER_MODE_BOSS:
+                events.Reset();
+                me->GetMotionMaster()->Clear();
+                break;
+            case BAT_RIDER_MODE_TRASH:
+                events.Reset();
+
+                // apply the Thrash (8876) spell to the bat rider (passive ability)
+                me->CastSpell(me, SPELL_PASSIVE_THRASH);
+
+                break;
+        }
     }
 
-    void JustEngagedWith(Unit* /*who*/) override
+    void JustEngagedWith(Unit* who) override
     {
-        _scheduler.Schedule(2s, [this](TaskContext context)
+        CreatureAI::JustEngagedWith(who);
+
+        switch (_mode)
         {
-            DoCastRandomTarget(SPELL_THROW_LIQUID_FIRE);
-            context.Repeat(7s);
-        });
+            case BAT_RIDER_MODE_BOSS:
+                events.ScheduleEvent(EVENT_BAT_RIDER_THROW_BOMB, 2s);
+                break;
+            case BAT_RIDER_MODE_TRASH:
+                events.ScheduleEvent(EVENT_DEMO_SHOUT, 1s);
+                events.ScheduleEvent(EVENT_BATTLE_COMMAND, 8s);
+                events.ScheduleEvent(EVENT_INFECTED_BITE, 6500ms);
+        }
     }
 
     void UpdateAI(uint32 diff) override
     {
-        if (!UpdateVictim())
-            return;
+        events.Update(diff);
 
-        _scheduler.Update(diff);
-    }
+        if (_mode == BAT_RIDER_MODE_BOSS)
+        {
+            // if the creature isn't moving, run the loop
+            if (!me->isMoving())
+            {
+                LOG_DEBUG("scripts.ai", "npc_batrider::UpdateAI: not moving, running loop");
+                // enable flying
+                me->SetDisableGravity(true);
+                // send the rider on its loop
+                me->GetMotionMaster()->MoveSplinePath(PATH_BAT_RIDER_LOOP);
+            }
+
+            // event handling
+            switch (events.ExecuteEvent())
+            {
+                case EVENT_BAT_RIDER_THROW_BOMB:
+                    LOG_DEBUG("scripts.ai", "npc_batrider::UpdateAI: EVENT_BAT_RIDER_THROW_BOMB");
+                    DoCastRandomTarget(SPELL_THROW_LIQUID_FIRE);
+                    events.ScheduleEvent(EVENT_BAT_RIDER_THROW_BOMB, 8s);
+                    break;
+                default:
+                    break;
+            }
+        }
+        else if (_mode == BAT_RIDER_MODE_TRASH)
+        {
+            if (!UpdateVictim())
+            {
+                return;
+            }
+
+            // don't interrupt casting
+            if (me->HasUnitState(UNIT_STATE_CASTING))
+            {
+                return;
+            }
 
-private:
-    TaskScheduler _scheduler;
+            // if health goes below 30%, cast Unstable Concotion
+            if (me->HealthBelowPct(30))
+            {
+                events.ScheduleEvent(EVENT_UNSTABLE_CONCOCTION, 0);
+            }
+
+            // event handling
+            switch (events.ExecuteEvent())
```

Comment:
why not create a private TaskScheduler and make the bat rider abilities fire with the taskscheduler as well?

Reply:
Done.